crates/transport/src/error.rs

```diff
@@ -67,4 +72,109 @@ impl TransportErrorKind {
     pub const fn pubsub_unavailable() -> TransportError {
         RpcError::Transport(Self::PubsubUnavailable)
     }
+
+    /// Instantiate a new `TrasnportError::HttpError`.
+    pub const fn http_error(status: u16, body: String) -> TransportError {
+        RpcError::Transport(Self::HttpError(HttpError { status, body }))
+    }
+
+    /// Analyzes the [TransportErrorKind] and decides if the request should be retried based on the
+    /// variant.
+    pub fn is_retry_err(&self) -> bool {
+        match self {
+            // Missing batch response errors can be retried.
+            Self::MissingBatchResponse(_) => true,
+            Self::HttpError(http_err) => http_err.is_rate_limit_err(),
+            Self::Custom(err) => {
+                let msg = err.to_string();
+                msg.contains("429 Too Many Requests")
+            }
+            // If the backend is gone, or there's a completely custom error, we should assume it's
+            // not retryable.
+            _ => false,
+        }
+    }
+}
+
+/// Type for holding HTTP errors such as 429 rate limit error.
+#[derive(Debug, thiserror::Error)]
+#[error("HTTP error {status} with body: {body}")]
+pub struct HttpError {
+    pub status: u16,
+    pub body: String,
+}
+
+impl HttpError {
+    /// Checks the `status` to determine whether the request should be retried.
+    pub const fn is_rate_limit_err(&self) -> bool {
+        if self.status == 429 {
+            return true;
+        }
+        false
+    }
+}
+
+/// Extension trait to implement methods for [`RpcError<TransportErrorKind, E>`].
+pub trait RpcErrorExt<E> {
```

Comment:
what's this generic for?

Reply:
good question, what is this for?
or in general, why do we need this?

Reply:
Yeah my bad! It isn't required. I thought E i.e `ErrorPayload<ErrData>` would need a Borrow bound.